src/transform/src/fusion/top_k.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Fuses a sequence of `TopK` operators in to one `TopK` operator
+
+use crate::TransformArgs;
+use expr::MirRelationExpr;
+
+/// Fuses a sequence of `TopK` operators in to one `TopK` operator if
+/// they happen to share the same grouping and ordering key.
+#[derive(Debug)]
+pub struct TopK;
+
+impl crate::Transform for TopK {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        relation.visit_mut_pre(&mut |e| {
+            self.action(e);
+        });
+        Ok(())
+    }
+}
+
+impl TopK {
+    /// Fuses a sequence of `TopK` operators in to one `TopK` operator.
+    pub fn action(&self, relation: &mut MirRelationExpr) {
+        if let MirRelationExpr::TopK {
+            input,
+            group_key,
+            order_key,
+            limit,
+            offset,
+            monotonic,
+        } = relation
+        {
+            while let MirRelationExpr::TopK {
+                input: inner_input,
+                group_key: inner_group_key,
+                order_key: inner_order_key,
+                limit: inner_limit,
+                offset: inner_offset,
+                monotonic: inner_monotonic,
+            } = &mut **input
+            {
+                if *group_key == *inner_group_key && *order_key == *inner_order_key {
+                    if let Some(inner_limit) = inner_limit {
+                        if *offset >= *inner_limit {
+                            relation.take_safely();
+                            break;
+                        }
+                        *inner_limit -= *offset;
+                        if let Some(limit) = limit {
+                            *limit = std::cmp::min(*limit, *inner_limit);
+                        } else {
+                            *limit = Some(*inner_limit);
+                        }
+                    }
+                    *offset += *inner_offset;
+                    *monotonic = *inner_monotonic;
+                    **input = inner_input.take_dangerous();
```

Comment:
I have a small preference for finding a way to explain this that is clear to someone without needing pencil and paper. If you have that (e.g. some nearby pencil and paper that you used) dropping it in as comments would be great! I'm not 100% confident I'm able to verify the logic.

Reply:
The way I intuit this that the inner topK means that the outer topK will receive at most `inner_limit` rows. The outer topK will skip the first `outer_offset` of the `inner_limit` rows and return the first `outer_limit` rows, if they exist.
```
 inner_offset  inner_limit
|------------|xxxxxxxxxxxxxxxxxx|
             |------------|xxxxxxxxxxxx|
              outer_offset   outer_limit
```
Thus, the combined offset is the `inner_offset + outer_offset`.
The combined limit is `min(max(inner_limit - outer_offset, 0), outer_limit)`. If a limit is None, then it should be treated as infinity.

Hmm... how about we
1) take out these lines:
```
                        if *offset >= *inner_limit {
                            relation.take_safely();
                            break;
                        }
```
2) Calculate limit according to the formula above.
3) Outside of this loop, call relation.take_safely() if `limit` is 0.

This would allow turning the expression into an empty constant if `outer_limit` is 0 as well as if `outer_offset >= inner_limit`.

Reply:
Much more clear that way. Thanks for the ASCII art too :)